watcher.go

```diff
@@ -0,0 +1,140 @@
+package consul
+
+import (
+	"context"
+	"net"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+// WatcherFunc is the function signature for the callback from watch.  It
+// passes a list of KeyData representing the most recent value or values stored
+// at the key or everything below the prefix. It can be nil.  error is passed
+// through to the handler if it is not temporary. Caller is responsible for
+// stopping the watch if error is determined to be unrecoverable If the watcher
+// encounters errors, it will retry up to MaxAttempts times before a successful
+// attempt or else it will bail entirely.  If the error is temporary, it will
+// not pass it through to the WatcherFunc unless MaxAttempts was hit
+type WatcherFunc func([]KeyData, error)
+
+// Watcher is the struct upon which Watch and WatchPrefix are built.
+type Watcher struct {
+	Client *Client
+	// MaxAttempts limits the number of subsequent failed API calls before
+	// bailing out of the watch.
+	MaxAttempts int
+}
+
+var (
+	DefaultWatcher = &Watcher{}
+
+	// WatchTransport is the same as DefaultTransport with a longer
+	// ResponseHeaderTimeout.  This is copied from DefaultTransport.  We don't
+	// do an actual copy because the Trasnport uses mutexes.  copied mutexes
+	// are ineffective.  You will panic on concurrent map access if running
+	// with other clients without instantiating a new transport for watch.
+	WatchTransport = &http.Transport{
```

Comment:
To follow the model of the DefaultTransport in both this package and the net/http package this should be of type `http.RoundTripper` so a program can override it to whatever transport it needs (setting one that logs or collects stats for example).

See https://github.com/segmentio/consul-go/blob/master/client.go#L33